# Final Project: Crypto Currency from 2013 to 2021

Member of our team:

<table style='float:center'>
    <tr>
        <th>Fullname</th>
        <th>Student ID</th>
    </tr>
    <tr>
        <td>Đoàn Ánh Dương</td>
        <td>20127474</td>
    </tr>
</table>




---

## Import Library

Necessary file imported here.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

import os
import glob

Environment using within this final project.

In [2]:
import sys
sys.executable

'/home/duong/anaconda3/envs/min_ds-env/bin/python'

Set options here.

In [3]:
pd.set_option('display.max_colwidth', None)
# pd.set_option('display.float_format', '{:.2f}'.format)

---

## I. Idea and Expectation
### Idea
For recent years, we have been heard at least once about crypto, particuarly Bitcoin and its application. Although digital currency brought many controversial opinions up to now, we cannot realize its success and popularity. 

So I decided to deep in crypto market and fortunately, I found the historical price data of crypto currency on Kaggle. Although the data does not contain all information about a specific coin, the historical price can bring us to take an overview of digital currency evolution.

Throughout this project, I hope that we can have a perspective objectively and a basic knowledge about digital currency.
### Expectation
In this project, we will find out the crypto currency market and its history, with expectation that we can find a method to buy a coin and take profit. 

## II. Datasets and Collection

### Source and Public Data
I found the datasets used in this project was public on [Kaggle](https://www.kaggle.com/datasets/sudalairajkumar/cryptocurrencypricehistory): Crypto Currrency price history.

### License
The data is taken from [coinmarketcap](https://coinmarketcap.com/) and it is [free](https://coinmarketcap.com/faq/) to use the data.

More about the license I found on kaggle: 
- CC0: Public Domain ([No copyright](https://creativecommons.org/publicdomain/zero/1.0/))

### About Data
The dataset has one csv file for each currency. **Price history is available on a daily basis from <font color = 'green'>*April 28, 2013*</font>**. This dataset has the historical price information of some of the top crypto currencies by market capitalization, including Bitcoin, Etherium, Binance Coin,...

### Description 
All information of columns contain in the file named `description.txt`.

### Author
Datasets were collected by Sudalai Rajkumar(registered account name [SRK](https://www.kaggle.com/sudalairajkumar)) on Kaggle.

### How did the author collect data?
As the author's answer, there is not mentioned clearly how the way he collected data, maybe collected by scraping historical data web. For instance: [bitcoin](https://coinmarketcap.com/currencies/bitcoin/historical-data/).

### Is dataset subjective or objective?
Besides the historical price of digital currency, there are many factors in a coin we need to take consideration on. For example: Holders Statistic over time, legality, team and organization investors, who created the coin... 

So that the dataset is subjective.

---

## III. Preprocessing Data

### 1. Change files.

First, we cannot access each files to analyze, it must be combined into one.

#### How many files does we have?

In [4]:
path = 'archive/'
os.chdir(path)
ext = 'csv'
all_files = [f for f in glob.glob(f'*.{ext}')]

print(f'Containing number of files: {len(all_files)}.')
print(all_files)

os.chdir('../')

Containing number of files: 23.
['coin_XRP.csv', 'coin_Ethereum.csv', 'coin_Aave.csv', 'coin_USDCoin.csv', 'coin_Solana.csv', 'coin_EOS.csv', 'coin_WrappedBitcoin.csv', 'coin_Cardano.csv', 'coin_NEM.csv', 'coin_Tron.csv', 'coin_CryptocomCoin.csv', 'coin_BinanceCoin.csv', 'coin_Iota.csv', 'coin_ChainLink.csv', 'coin_Monero.csv', 'coin_Dogecoin.csv', 'coin_Tether.csv', 'coin_Bitcoin.csv', 'coin_Cosmos.csv', 'coin_Litecoin.csv', 'coin_Stellar.csv', 'coin_Polkadot.csv', 'coin_Uniswap.csv']


#### Are there duplicated in each files? 

Lets check duplicated rows.

In [5]:
for f in all_files:
    temp = pd.read_csv(path+f)
    print(f'duplicated in {f}:', temp.index.duplicated().sum())

duplicated in coin_XRP.csv: 0
duplicated in coin_Ethereum.csv: 0
duplicated in coin_Aave.csv: 0
duplicated in coin_USDCoin.csv: 0
duplicated in coin_Solana.csv: 0
duplicated in coin_EOS.csv: 0
duplicated in coin_WrappedBitcoin.csv: 0
duplicated in coin_Cardano.csv: 0
duplicated in coin_NEM.csv: 0
duplicated in coin_Tron.csv: 0
duplicated in coin_CryptocomCoin.csv: 0
duplicated in coin_BinanceCoin.csv: 0
duplicated in coin_Iota.csv: 0
duplicated in coin_ChainLink.csv: 0
duplicated in coin_Monero.csv: 0
duplicated in coin_Dogecoin.csv: 0
duplicated in coin_Tether.csv: 0
duplicated in coin_Bitcoin.csv: 0
duplicated in coin_Cosmos.csv: 0
duplicated in coin_Litecoin.csv: 0
duplicated in coin_Stellar.csv: 0
duplicated in coin_Polkadot.csv: 0
duplicated in coin_Uniswap.csv: 0


So there is no duplicated rows. We carry out combining files.

In [6]:
combined_csv = pd.concat([pd.read_csv(path + fname) for fname in all_files])
print('duplicated in combine files: ', combined_csv.index.duplicated().sum())

duplicated in combine files:  34091


Why the combined files take duplicated? Because column 'SNo' is the index starting with $1$ for each files. So we must reset the index in combined files. 

In [7]:
combined_csv.drop('SNo',1,inplace = True)
combined_csv.to_csv('crypto_finance.csv', index=False)

combined_csv = pd.read_csv('crypto_finance.csv')
combined_csv.index.rename('SNo', inplace = True)
combined_csv.to_csv('crypto_finance.csv')

### 2. Pre-processing

Read new file. Let's see some values.

In [8]:
cryp_df = pd.read_csv('crypto_finance.csv', index_col='SNo')

cryp_df.sample(10)

Name Symbol                 Date        High         Low  \
SNo                                                                          
1001               XRP    XRP  2016-05-02 23:59:59    0.006803    0.006712   
3666          Ethereum    ETH  2017-09-19 23:59:59  295.768005  274.484009   
34469          Stellar    XLM  2016-01-14 23:59:59    0.001914    0.001762   
6468            Solana    SOL  2020-08-27 23:59:59    3.854706    3.292214   
35296          Stellar    XLM  2018-04-20 23:59:59    0.398048    0.356500   
10254          Cardano    ADA  2020-10-24 23:59:59    0.109662    0.107539   
14892  Crypto.com Coin    CRO  2020-11-16 23:59:59    0.066765    0.063283   
11504              NEM    XEM  2017-12-21 23:59:59    1.051120    0.913511   
16943             IOTA  MIOTA  2018-06-25 23:59:59    1.046840    0.960835   
33205         Litecoin    LTC  2019-06-30 23:59:59  136.259679  120.440985   

             Open       Close        Volume     Marketcap  
SNo                                                        
1001     0.006791    0.006714  5.858260e+05  2.341121e+08  
3666   293.656006  282.803986  6.469090e+08  2.678709e+10  
34469    0.001824    0.001914  6.412650e+04  9.258701e+06  
6468     3.557844    3.445060  2.103136e+07  1.110603e+08  
35296    0.366551    0.391165  1.702820e+08  7.263791e+09  
10254    0.108083    0.108075  8.681945e+08  3.362486e+09  
14892    0.063953    0.064691  4.315723e+07  1.404831e+09  
11504    0.949966    0.998102  1.128350e+08  8.982918e+09  
16943    0.978272    1.018500  5.641030e+07  2.830952e+09  
33205  133.724642  122.249118  5.525273e+09  7.636047e+09

More information about dataset.

In [9]:
cryp_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 37082 entries, 0 to 37081
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Name       37082 non-null  object 
 1   Symbol     37082 non-null  object 
 2   Date       37082 non-null  object 
 3   High       37082 non-null  float64
 4   Low        37082 non-null  float64
 5   Open       37082 non-null  float64
 6   Close      37082 non-null  float64
 7   Volume     37082 non-null  float64
 8   Marketcap  37082 non-null  float64
dtypes: float64(6), object(3)
memory usage: 2.8+ MB


#### How many rows and columns in the file?

In [10]:
cryp_df.shape

(37082, 9)

There are 37082 rows and 9 columns.

#### What is the meaning of each row (sample)?

The dataset shows that each samples contain the information of a specific coin at each time.

#### Does the data have duplicated samples?

In [11]:
cryp_df.index.duplicated().sum()

0

There is no duplicated samples.

#### What does each column mean? 

In [12]:
f = open('description.txt', 'r')
print(f.read())
f.close()

1) SNo:         Serial Number.
2) Name:        Name of coin.
3) Symbol:      Symbol of coin.
4) Date:        Date of observation.
5) Open:        Opening price on the given day.
6) High:        Highest price on the given day.
7) Low:         Lowest price on the given day.
8) Close:       Closing price on the given day.
9) Volume:      Volume of transactions on the given day (in USD).
10) Market Cap: Market capitalization in USD.


#### Check the data types? 

In [13]:
cryp_df.dtypes

Name          object
Symbol        object
Date          object
High         float64
Low          float64
Open         float64
Close        float64
Volume       float64
Marketcap    float64
dtype: object

Overall, all the column types is correct except `Date`. We need to change `Date` column to appropriate type (`datetime`).

In [14]:
cryp_df['Date'] = pd.to_datetime(cryp_df['Date'], format="%Y-%m-%d %X")

#### With the numeric columns, how are values distributed? Is there any abnormal value?

In [15]:
nume_cols_df = cryp_df.describe(include=['number', 'datetime'], datetime_is_numeric=True)
nume_cols_df.loc['missing_ratio'] = cryp_df.agg(lambda x: x.isna().mean()*100)

nume_cols_df

Date          High           Low  \
count                                  37082  37082.000000  37082.000000   
mean           2018-08-16 07:12:30.221616896   1016.058015    952.987707   
min                      2013-04-29 23:59:59      0.000089      0.000079   
25%                      2017-03-05 23:59:59      0.075664      0.069536   
50%                      2019-01-09 23:59:59      1.008733      0.999850   
75%                      2020-05-13 23:59:59     31.916399     28.996246   
max                      2021-07-06 23:59:59  64863.098908  62208.964366   
std                                      NaN   5249.503670   4907.932082   
missing_ratio                            0.0      0.000000      0.000000   

                       Open         Close        Volume     Marketcap  
count          37082.000000  37082.000000  3.708200e+04  3.708200e+04  
mean             985.323755    987.120511  3.022542e+09  1.542943e+10  
min                0.000086      0.000086  0.000000e+00  0.000000e+00  
25%                0.072456      0.072648  4.937190e+06  2.395955e+08  
50%                1.001157      1.001138  8.512805e+07  1.405335e+09  
75%               30.459673     30.512205  9.388489e+08  5.159305e+09  
max            63523.754869  63503.457930  3.509679e+11  1.186364e+12  
std             5088.101367   5093.703878  1.190963e+10  7.059128e+10  
missing_ratio      0.000000      0.000000  0.000000e+00  0.000000e+00

There is no missing value in DataFrame. But it exists abnormal values in `min` row of `Volumn` and `Marketcap` col.

So, lets take a look at these columns.

In [21]:
check_df = cryp_df[(cryp_df['Marketcap'] == 0) | (cryp_df['Volume'] == 0)]
check_df

Name Symbol                Date      High       Low      Open  \
SNo                                                                        
0           XRP    XRP 2013-08-05 23:59:59  0.005980  0.005613  0.005875   
1           XRP    XRP 2013-08-06 23:59:59  0.005661  0.004629  0.005637   
2           XRP    XRP 2013-08-07 23:59:59  0.004682  0.004333  0.004669   
3           XRP    XRP 2013-08-08 23:59:59  0.004424  0.004175  0.004397   
4           XRP    XRP 2013-08-09 23:59:59  0.004367  0.004253  0.004257   
...         ...    ...                 ...       ...       ...       ...   
36477  Polkadot    DOT 2020-08-28 23:59:59  6.333746  5.540963  5.639486   
36478  Polkadot    DOT 2020-08-29 23:59:59  6.562906  6.042309  6.175925   
36479  Polkadot    DOT 2020-08-30 23:59:59  6.219506  5.749978  6.153440   
36480  Polkadot    DOT 2020-08-31 23:59:59  6.459377  5.772966  5.905918   
36481  Polkadot    DOT 2020-09-01 23:59:59  6.838898  6.172324  6.298810   

          Close        Volume     Marketcap  
SNo                                          
0      0.005613  0.000000e+00  4.387916e+07  
1      0.004680  0.000000e+00  3.659101e+07  
2      0.004417  0.000000e+00  3.453412e+07  
3      0.004254  0.000000e+00  3.325863e+07  
4      0.004291  0.000000e+00  3.354750e+07  
...         ...           ...           ...  
36477  6.159955  7.271622e+08  0.000000e+00  
36478  6.159143  5.272900e+08  0.000000e+00  
36479  5.869881  4.853351e+08  0.000000e+00  
36480  6.300020  5.126048e+08  0.000000e+00  
36481  6.288767  6.015274e+08  0.000000e+00  

[971 rows x 9 columns]

We got some problems here: Abnormal values on `Volume` and `Marketcap` col. We can calculate the ratio of abnormal values to decide whether to drop or not.

In [34]:
abn_df = check_df.groupby(['Symbol']).size()

all_df = cryp_df.groupby(['Symbol']).size()

(abn_df/all_df *100).dropna()

Symbol
AAVE     0.363636
ATOM     5.443787
BTC      8.090939
CRO      0.213904
DOGE     0.398551
DOT      3.750000
EOS      0.068213
LTC      8.090939
SOL     11.504425
TRX      1.005747
USDC     0.798403
WBTC    22.072072
XRP      4.977532
dtype: float64

As the results, the number of invalid values is quite small, we can delete samples containing invalid values. But once we dropped it, we will face the intermittence of time data. So I decided not to drop.

#### With the categorical columns, how are the values distributed? Is there any abnormal value? 

In [36]:
cate_cols = cryp_df.describe(include=[object]).columns

cate_cols_df = cryp_df[cate_cols].agg([
    lambda x: x.isna().mean() *100, 
    pd.Series.nunique, 
    pd.Series.unique])

cate_cols_df.rename(index ={
    '<lambda>': 'missing_ratio',
    'nunique':'num_dif_val',
    'unique':'diff_vals'
})

Name  \
missing_ratio                                                                                                                                                                                                                 0.0   
num_dif_val                                                                                                                                                                                                                    23   
diff_vals      [XRP, Ethereum, Aave, USD Coin, Solana, EOS, Wrapped Bitcoin, Cardano, NEM, TRON, Crypto.com Coin, Binance Coin, IOTA, Chainlink, Monero, Dogecoin, Tether, Bitcoin, Cosmos, Litecoin, Stellar, Polkadot, Uniswap]   

                                                                                                                                     Symbol  
missing_ratio                                                                                                                           0.0  
num_dif_val                                                                                                                              23  
diff_vals      [XRP, ETH, AAVE, USDC, SOL, EOS, WBTC, ADA, XEM, TRX, CRO, BNB, MIOTA, LINK, XMR, DOGE, USDT, BTC, ATOM, LTC, XLM, DOT, UNI]

There is no missing value in categorical columns. Everything seems to be fine.

## IV. Data Exploratory and Analysis. 

### Propose meaningful question & answer.

After preprocessing, we had an overview about the dataset. Lets turn back to our purpose. There are 2 main questions we must answer:
- How does the price fluctuations of currencies correlate with each other?
- Which coins do we need to take consideration?


### 1. How does the price fluctuations of currencies correlate with each other?

With this question, we could know how the digital currency work. The answer of this question tells us we should invest alternative coins or bitcoin.